homeassistant/components/device_tracker/cppm_tracker.py

```diff
@@ -0,0 +1,88 @@
+"""
+Support for ClearPass Policy Manager.
+
+Allows tracking devices with CPPM.
+"""
+import logging
+from datetime import timedelta
+import voluptuous as vol
+from homeassistant.util import Throttle
+import homeassistant.helpers.config_validation as cv
+from homeassistant.components.device_tracker import (
+    PLATFORM_SCHEMA, DeviceScanner, DOMAIN
+)
+from homeassistant.const import (
+    CONF_HOST, CONF_API_KEY, CONF_SCAN_INTERVAL
+)
+
+REQUIREMENTS = ['clearpasspy==1.0.2']
+
+SCAN_INTERVAL = timedelta(seconds=120)
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_HOST): cv.string,
+    vol.Required('client_id'): cv.string,
+    vol.Required(CONF_API_KEY): cv.string,
+})
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def get_scanner(hass, config):
+    """Initialize Scanner."""
+    grant_type = 'client_credentials'
+    new_scan = CPPMDeviceScanner(config[DOMAIN], grant_type)
+    return new_scan if new_scan.success_init else None
+
+
+class CPPMDeviceScanner(DeviceScanner):
+    """Initialize class."""
+
+    def __init__(self, config, grant_type):
+        """Initialize class."""
+        _LOGGER.debug("-------------INIT CALLED--------------")
+        self._cppm_host = config[CONF_HOST]
+        self._api_key = config[CONF_API_KEY]
+        self._grant_type = grant_type
+        self._client_id = config['client_id']
+        self.success_init = self.get_cppm_data()
```

Comment:
Please move this out of init method. We don't want side effects in init method.

Reply:
Do you mean set `self.success_init` outside of init method?

Reply:
Call `get_cppm_data` in `get_scanner` and return the scanner if ok, if not return None.

`success_init` isn't really needed.

Reply:
As you can probably tell, Python is not my first language. I was working from other Device Tracker components that seem to do it the same way. Is this better?
```
def get_scanner(hass, config):
    """Initialize Scanner."""
    new_scan = CPPMDeviceScanner(config[DOMAIN], GRANT_TYPE)
    return new_scan if new_scan.access_token is not None else None
```
. . .
```
    def __init__(self, config, grant_type):
        """Initialize class."""
        from clearpasspy import ClearPass

        data = {
            'server': config[CONF_HOST],
            'grant_type': grant_type,
            'secret': config[CONF_API_KEY],
            'client': config[CLIENT_ID]
        }
        self.cppm = ClearPass(data)
        self.access_token = self.cppm.access_token
        #self.success_init = self.get_cppm_data()
```

Reply:
No we don't want any side effects in init method. Create the cppm instance first and check that it can talk to the api. If it can't talk to the api, return `None` from `get_scanner`. If it can talk to the api, create the scanner instance, passing the cppm instance to the scanner. Then return the scanner instance from `get_scanner`.

Reply:
So more or less, move everything from init to get_scanner?

```
def get_scanner(hass, config):
    """Initialize Scanner."""
    from clearpasspy import Clearpass
    data = {
        'server': config[CONF_HOST],
        'grant_type': grant_type,
        'secret': config[CONF_API_KEY],
        'client': config[CLIENT_ID]
    }
    cppm = ClearPass(data)
    if cppm.access_token is not None:
        new_scan = CPPMDeviceScanner(config[DOMAIN], GRANT_TYPE)
    else:
        return None
```

Reply:
Yeah, but I would invert the check and make it a guard clause.
```py
if cppm.access_token is None:
    return None
return CPPMDeviceScanner(config[DOMAIN], cppm)
```

Reply:
I will make the adjustment. Out of curiousity: What do you mean by guard clause?